From 512f7ea331af8fbb2076681a806b4a49077679b2 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 16:40:04 -0400
Subject: [PATCH] slub: add basic full slab sanitization

This is a partial port of the feature from PaX without the default fast
mode based on passing SLAB_NO_SANITIZE in performance-critical cases
that are not particularly security sensitive.

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 init/Kconfig | 10 ++++++++++
 mm/slub.c    | 20 ++++++++++++++++++++
 2 files changed, 30 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index 38eae61bad9c..03a581cb7340 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2060,6 +2060,16 @@ config SLAB_HARDENED
           Make minor performance sacrifices to harden the kernel slab
           allocator.
 
+config SLAB_SANITIZE
+	bool "Sanitize SLAB allocations"
+	depends on SLUB
+	default y
+	help
+	  Zero fill slab allocations on free, reducing the lifetime of
+	  sensitive data and helping to mitigate use-after-free bugs.
+
+	  For slabs with debug poisoning enabling, this has no impact.
+
 config SLUB_CPU_PARTIAL
 	default y
 	depends on SLUB && SMP
diff --git a/mm/slub.c b/mm/slub.c
index 35b097b7b9e9..1ef15f2a68e6 100644
--- a/mm/slub.c
+++ b/mm/slub.c
@@ -125,6 +125,11 @@ static inline int kmem_cache_debug(struct kmem_cache *s)
 #endif
 }
 
+static inline bool has_sanitize(struct kmem_cache *s)
+{
+	return IS_ENABLED(CONFIG_SLAB_SANITIZE) && !(s->flags & (SLAB_DESTROY_BY_RCU | SLAB_POISON));
+}
+
 void *fixup_red_left(struct kmem_cache *s, void *p)
 {
 	if (kmem_cache_debug(s) && s->flags & SLAB_RED_ZONE)
@@ -2963,6 +2968,21 @@ static __always_inline void do_slab_free(struct kmem_cache *s,
 	void *tail_obj = tail ? : head;
 	struct kmem_cache_cpu *c;
 	unsigned long tid;
+
+	if (has_sanitize(s)) {
+		int offset = s->offset ? 0 : sizeof(void *);
+		void *x = head;
+
+		while (1) {
+			memset(x + offset, 0, s->object_size - offset);
+			if (s->ctor)
+				s->ctor(x);
+			if (x == tail_obj)
+				break;
+			x = get_freepointer(s, x);
+		}
+	}
+
 redo:
 	/*
 	 * Determine the currently cpus per cpu slab.
